### PR TITLE
fix: directory traversal vulnerability

### LINKED ISSUE #3016: Directory traversal bug
My team found a weakness that could open up security vulnerabilities in the right circumstances. It happens when un-trusted users are allowed pass params that are templated into input artifact destination paths. It becomes worse when un-trusted users are able to control the source of the artifact and/or the step is designed in a way that non-executable files are made executable.

For example, if I allow a user to place a file on S3 that is then pulled in and written to path: /tmp/{{inputs.parameters.filename}}, that user can write/overwrite arbitrary files anywhere on the container using directory traversal (..).

If the docker bootstraps a non-executable file to run (e.g. java {classname}), I can replace that file (e.g. /app/classes/SomeApp.class) with a malicious file.Obviously that's a lot of things that have to go wrong before there's a workable exploit, but I'd argue they should be mitigated in a couple ways:

* directory traversal shouldn't be permitted in input artifact destination paths
* input artifacts shouldn't be allowed to overwrite files already on the filesystem

### PR DESCRIPTION
**Pull request by kennytrytek**
Fixes #3016

I'm not sure if this is all that is needed to fully resolve this issue, so please let me know if I've missed something.

Testing: ✅ 
Automated tests cover it since the path contents only matter just before it's used to read or write the file, and the artifact path is resolved before this point in the code.

Signed-off-by: Kenny Trytek <kenneth.g.trytek@gmail.com>

### CODE REVIEW

**Comment by kennytrytek** [comment]
I know lint is failing. I am unable to run the `pre-commit` Make target, and I'm not sure what commands will lint this file correctly. See output in collapsed comment below.
<details><summary>`make pre-commit -B`</summary>
<p>

```
$ make pre-commit -B
GIT_COMMIT=b9736d95bd5a78bdddea96ee8149a94d4958a86d GIT_BRANCH=kgt/directory-traversal-3016 GIT_TAG=untagged GIT_TREE_STATE=clean RELEASE_TAG=false DEV_BRANCH=true VERSION=latest
KUBECTX=local DOCKER_DESKTOP=false K3D=false DOCKER_PUSH=false
RUN_MODE=local PROFILE=minimal AUTH_MODE=hybrid SECURE=false STATIC_FILES=false ALWAYS_OFFLOAD_NODE_STATUS=false UPPERIO_DB_DEBUG=0 LOG_LEVEL=debug NAMESPACED=true
/Users/kennytrytek/go/bin/go-to-protobuf \
	--go-header-file=./hack/custom-boilerplate.go.txt \
	--packages=github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1 \
	--apimachinery-packages=+k8s.io/apimachinery/pkg/util/intstr,+k8s.io/apimachinery/pkg/api/resource,k8s.io/apimachinery/pkg/runtime/schema,+k8s.io/apimachinery/pkg/runtime,k8s.io/apimachinery/pkg/apis/meta/v1,k8s.io/api/core/v1,k8s.io/api/policy/v1beta1 \
	--proto-import /Users/kennytrytek/go/src
topological order bytes
topological order encoding/base64
topological order encoding/json
topological order errors
topological order fmt
topological order github.com/argoproj/argo-workflows/v3/errors
topological order github.com/argoproj/argo-workflows/v3/pkg/apis/workflow
topological order github.com/argoproj/argo-workflows/v3/util/json
topological order github.com/argoproj/argo-workflows/v3/util/slice
topological order github.com/go-openapi/spec
topological order github.com/gogo/protobuf/proto
topological order github.com/gogo/protobuf/sortkeys
topological order github.com/google/gofuzz
topological order go/ast
topological order go/doc
topological order go/parser
topological order go/token
topological order gopkg.in/inf.v0
topological order hash/fnv
topological order io
topological order io/ioutil
topological order k8s.io/apimachinery/pkg/conversion
topological order k8s.io/apimachinery/pkg/conversion/queryparams
topological order k8s.io/apimachinery/pkg/fields
topological order k8s.io/apimachinery/pkg/labels
topological order k8s.io/apimachinery/pkg/selection
topological order k8s.io/apimachinery/pkg/types
topological order k8s.io/apimachinery/pkg/util/errors
topological order k8s.io/apimachinery/pkg/util/json
topological order k8s.io/apimachinery/pkg/util/naming
topological order k8s.io/apimachinery/pkg/util/runtime
topological order k8s.io/apimachinery/pkg/util/sets
topological order k8s.io/apimachinery/pkg/util/validation
topological order k8s.io/apimachinery/pkg/watch
topological order k8s.io/klog/v2
topological order k8s.io/kube-openapi/pkg/common
topological order math
topological order math/big
topological order math/bits
topological order net/url
topological order os
topological order path
topological order path/filepath
topological order reflect
topological order regexp
topological order runtime/debug
topological order sigs.k8s.io/structured-merge-diff/v4/value
topological order sigs.k8s.io/yaml
topological order sort
topological order strconv
topological order strings
topological order sync
topological order sync/atomic
topological order time
topological order unsafe
topological order k8s.io/apimachinery/pkg/runtime/schema
topological order k8s.io/apimachinery/pkg/util/intstr
topological order k8s.io/apimachinery/pkg/api/resource
topological order k8s.io/apimachinery/pkg/runtime
topological order k8s.io/apimachinery/pkg/apis/meta/v1
topological order k8s.io/api/core/v1
topological order k8s.io/api/policy/v1beta1
topological order github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1
2021/11/08 20:20:06 k8s.io/apimachinery/pkg/runtime/generated.proto: File not found.
k8s.io/apimachinery/pkg/runtime/schema/generated.proto: File not found.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:9:1: Import "k8s.io/apimachinery/pkg/runtime/generated.proto" was not found or had errors.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:10:1: Import "k8s.io/apimachinery/pkg/runtime/schema/generated.proto" was not found or had errors.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:447:12: "k8s.io.apimachinery.pkg.runtime.RawExtension" is not defined.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:1146:12: "k8s.io.apimachinery.pkg.runtime.RawExtension" is not defined.
2021/11/08 20:20:06 protoc -I . -I /Users/kennytrytek/go/src -I /Users/kennytrytek/go/src --gogo_out=/Users/kennytrytek/go/src /Users/kennytrytek/go/src/github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto
2021/11/08 20:20:06 Unable to generate protoc on k8s.io.apimachinery.pkg.apis.meta.v1: exit status 1
Makefile:300: *** [pkg/apis/workflow/v1alpha1/generated.proto] error 1

#0  pkg/apis/workflow/v1alpha1/generated.proto at /Users/kennytrytek/go/src/github.com/argoproj/argo-workflows/Makefile:299
#1  types at /Users/kennytrytek/go/src/github.com/argoproj/argo-workflows/Makefile:246
#2  codegen at /Users/kennytrytek/go/src/github.com/argoproj/argo-workflows/Makefile:241
#3  pre-commit at /Users/kennytrytek/go/src/github.com/argoproj/argo-workflows/Makefile:570
```

</p>
</details>

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -840,6 +843,17 @@ type Artifact struct {
 	FromExpression string `json:"fromExpression,omitempty" protobuf:"bytes,11,opt,name=fromExpression"`
 }
 
+// ValidatePath verifies that the artifact path is valid.
+func (art Artifact) ValidatePath() error {
+	if art.Path == "" {
+		return argoerrs.InternalErrorf("Artifact %s did not specify a path", art.Name)
+	}
+	if strings.Contains(art.Path, "..") {
```
Comment (crenshaw-dev) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:851]: I think this check should be more robust. As far as I can tell, `a..a` is a valid filename in *nix.

Will look into whether golang has a library to detect directory traversal...

File: pkg/apis/workflow/v1alpha1/workflow_types_test.go [hunk 2]
```diff
@@ -79,6 +79,23 @@ func TestWorkflowHappenedBetween(t *testing.T) {
...
+	t.Run("directory traversal fails", func(t *testing.T) {
+		err := Artifact{Name: "a1", Path: "../../etc/passwd"}.ValidatePath()
+		assert.Contains(t, err.Error(), "Directory traversal is not permitted")
+	})
+
+	t.Run("normal artifact path succeeds", func(t *testing.T) {
+		err := Artifact{Name: "a1", Path: "/tmp/some-artifact.txt"}.ValidatePath()
+		assert.NoError(t, err)
+	})
+}
```
Comment (crenshaw-dev) [inline pkg/apis/workflow/v1alpha1/workflow_types_test.go:97]: Add a test to make sure it allows filenames with embedded `..`.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 1 above)
Comment (crenshaw-dev) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:849]: I guess once could use [`filepath.Clean`](https://pkg.go.dev/path/filepath#Clean) and then check that the generated path begins with the full root path of the artifacts directory.

Comment (kennytrytek) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:849]: One option would be to fully resolve the path with `path.Clean` and make sure the final path is within some safe boundary, but I'm not sure what the safe boundary would be. Checking for `../` doesn't seem to be sufficient either since `a..` is a valid file name.

Comment (crenshaw-dev) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:849]: Here's one idea: 

```go
func isSafePath(pathToTest string) bool {
	// Convert the given path to an absolute path by prepending "/" (not cross-platform?)
	cleanPath := path.Clean("/" + pathToTest)
	// Assert that the given path is no different from the clean path (i.e. no traversal)
	return cleanPath == "/" + pathToTest
}

func main() {
	fmt.Println(isSafePath("../whatever"))
	fmt.Println(isSafePath("whatever"))
}
```

Downsides:
* probably not cross-platform
* doesn't allow directory traversal _within_ the safe area (which would be nice, but probably not must-have)

**Review by simster7 — COMMENTED** [review]
This is one of those things that seems simple in the outset, but has a ton of edge cases. I had a rough time implementing this for Argo CD a while back, it took me four tries to get it right:

https://github.com/argoproj/argo-cd/pull/2452 -> https://github.com/argoproj/argo-cd/pull/2716 -> https://github.com/argoproj/argo-cd/pull/2797 -> https://github.com/argoproj/argo-cd/pull/2821

Take a look at the last PR in that chain to see what stuck and the attached test cases. Main learning: do not rely on string/regex verification and use Go's filepath utilities.

Using any sort of string regex or pure string validation is not sufficient and can lead to edge cases. For example, consider we are in `/var/dir1`, the path `../dir1/sub-dir` is valid even though it starts with `..`.

**Comment by kennytrytek** [comment]
>@simster7:
>Using any sort of string regex or pure string validation is not sufficient

I think in this case, all the possibilities can be enumerated, making a string manipulation solution possible. Since this is specifically preventing directory traversal in the container artifact mount path, there are only a few possibilities.

|#|Template|User Input|Raw Path|Cleaned Path|Behavior Needs to Change|
|---|---|---|---|---|---|
|1|{{inputs.parameters.path}}|..|..|/|No|
|2|{{inputs.parameters.path}}|../|../|/|No|
|3|{{inputs.parameters.path}}/tmp|..|../tmp|/tmp|No|
|4|/tmp/{{inputs.parameters.path}}|..|/tmp/..|/|Yes|
|5|/tmp/{{inputs.parameters.path}}|../abc|/tmp/../abc|/abc|Yes|
|6|/tmp/{{inputs.parameters.path}}|../tmp/abc|/tmp/../tmp/abc|/tmp/abc|No|
|7|/tmp/{{inputs.parameters.path}}|abc|/tmp/abc|/tmp/abc|No|

For examples 1, 2, and 3, I don't think there's anything we can do since the template creator has already shot themselves in the foot by allowing untrusted user input to define the path anywhere in the container.

Examples 4 and 5 are the problematic ones, and can be resolved by:
* parsing the input string up to the point of the first directory traversal (Regex `/\.\.$` or `/\.\./`) to find the "safe" directory.
* using `filepath.Clean` to remove all directory traversals.
* validating that the cleaned filepath lands within the safe directory.

Example 6 will be resolved by the same process as examples 4 and 5, even though it is not problematic in the same way.

Example 7 is the normal case, and will continue to function as usual.

If I have missed anything, let me know. I think this approach will be sufficient, provide I use the os-agnostic path separators available in Go instead of hard-coding `/` or `\\`.

File: pkg/apis/workflow/v1alpha1/workflow_types_test.go (hunk 2 above)
Comment (kennytrytek) [reply pkg/apis/workflow/v1alpha1/workflow_types_test.go:189]: Added in https://github.com/argoproj/argo-workflows/pull/7187/commits/6fed6857743cd4abeef3951fe4e34485162ea5f3

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 1 above)
Comment (kennytrytek) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: Made this function more robust in https://github.com/argoproj/argo-workflows/pull/7187/commits/6fed6857743cd4abeef3951fe4e34485162ea5f3. See test cases for some clear examples of how it behaves.

File: workflow/executor/executor.go [hunk 3]
```diff
@@ -169,6 +169,10 @@ func (we *WorkflowExecutor) LoadArtifacts(ctx context.Context) error {
 				return argoerrs.Errorf(argoerrs.CodeNotFound, "required artifact '%s' not supplied", art.Name)
 			}
 		}
+		err := art.CleanPath()
```
Comment (kennytrytek) [inline workflow/executor/executor.go:172]: I moved this check ahead of the call to `newDriverArt` since `newDriverArt` makes a deep clone of this artifact, so I wanted it to be operating on an artifact that has a clean path.

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 4]
```diff
@@ -853,6 +857,37 @@ type Artifact struct {
...
+
+	// Ensure that the artifact path does not use directory traversal to escape a
+	// "safe" sub-directory, assuming malicious user input is present. For example:
+	// inputs:
+	//    artifacts:
+	//      - name: a1
+	//        path: /tmp/safe/{{ inputs.parameters.user-input }}
+	//
+	// Any resolved path should always be within the /tmp/safe/ directory.
+	safeDir := ""
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:874]: We should not use regex for this, they are very brittle, and sometimes create un-fixable bugs. This is demonstrated by the many edge case bug we have when they were used in the past.

Could we use `path.Clean`?

https://dzx.cz/2021/04/02/go_path_traversal/

**Review by alexec — DISMISSED** [review]
see comment

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 4 above)
Comment (kennytrytek) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:874]: Yes, `path.Clean` is used a few lines below. The "safeDir" is the best-effort guess as to what the template creator was attempting to do. See [this comment](https://github.com/argoproj/argo-workflows/pull/7187#issuecomment-970415218) above for a table of use cases, and the full suite of [test cases](https://github.com/argoproj/argo-workflows/pull/7187/files#diff-ebcb4d1a2cd98dee1d7ec51113c3dee40128e21e2d259699163449f34c431e98R83) that precisely show how this will behave.

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:874]: I think we should probably be cleaning at the point of usage, not before. I.e. before we invoke LoadArtifact and SaveArtifact. Thoughts?

Comment (kennytrytek) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:874]: The only usages outside `LoadArtifact` and `SaveArtifact` are in `workflowpod.addInputArtifactsVolumes` and `validate.validateInputs`. In those places, there were previously errors raised if the artifact path was empty, so I thought it would also be prudent to further validate the given path by way of cleaning it.

Since `path.Clean` does not mutate the final destination from an OS perspective (i.e. `/tmp/../tmp` == `/tmp`), cleaning isn't strictly necessary. This method could be renamed `ValidatePath`, and remain just as useful.

I don't see any harm in this implementation as-is, since `path.Clean` should be idempotent. I'm open to other opinions, if someone has a better suggestion.

**Comment by kennytrytek** [comment]
Would anyone be able to review this? I believe I've addressed all the questions and comments.

**Comment by kennytrytek** [comment]
@alexec, any updates here?

**Comment by kennytrytek** [comment]
I am unable to change the "requested changes" state of this PR. If someone is available, I believe this is ready for review.

**Review by alexec — CHANGES_REQUESTED** [review]
My opinion on this is the same really.

We should not be using regular expressions for this task as in similar situations again, they've proven to be buggy, requiring constant attention.

Can we use `path.Clean` or some existing library function?

**Comment by kennytrytek** [comment]
>Can we use path.Clean or some existing library function?

That question was addressed in the earlier conversation on this PR:
https://github.com/argoproj/argo-workflows/pull/7187#discussion_r758798884

The short story is that this implementation is already using `path.Clean`. The regex is used to support guessing what the user was intending to be the "safe" path.

**Comment by kennytrytek** [comment]
This is not stale. Waiting on @alexec.

**Review by alexec — COMMENTED** [review]
I've been over this again today. We want to prevent directory traversal when a template is owned by the operator by the parameters come from the user. 

Correct me if I'm wrong, but rather than clean the path, we could return an error if the path contains `../`? We could feature flag it.

This would be simpler.

**Comment by kennytrytek** [comment]
> Correct me if I'm wrong, but rather than clean the path, we could return an error if the path contains `../`? We could feature flag it.

@alexec, checking for `../` is insufficient for validating whether an unsafe directory traversal has occurred. This was brought up by @crenshaw-dev in [this comment](https://github.com/argoproj/argo-workflows/pull/7187#discussion_r745807506), @simster7 in [this comment](https://github.com/argoproj/argo-workflows/pull/7187#pullrequestreview-803103114), and myself in [this comment](https://github.com/argoproj/argo-workflows/pull/7187#issuecomment-970415218). Case 9 in the table below gives a counter-example to that tactic.

A feature flag is not something I had considered, but if that would make it possible to deliver this fix instead of getting stuck for months, I'm all for it.

Here is a table showing the cases that this code is tested for. If any of these cases does not make sense, or if I'm missing a case, please let me know. These cases are validated under the assumption that the workflow template creator does not hard-code ".." in the workflow template, and that any ".." in the path must come from the user.

|#|Template|`user-input`|Raw Path|Cleaned Path|Raises Error|
|---|---|---|---|---|---|
|1|/tmp| |/tmp|/tmp|No|
|2|/tmp/| |/tmp/|/tmp|No|
|3|/tmp/abcd/some-artifact.txt| |/tmp/abcd/some-artifact.txt|/tmp/abcd/some-artifact.txt|No|
|4|{{user-input}}| | | |Yes|
|5|{{user-input}}|..|..|..|No|
|6|{{user-input}}|../|../|..|No|
|7|{{user-input}}|../..|../..|../..|No|
|8|{{user-input}}|../etc/passwd|../etc/passwd|../etc/passwd|No|
|9|/tmp/{{user-input}}|..|/tmp/..|/|Yes|
|10|/tmp/{{user-input}}|../|/tmp/../|/|Yes|
|11|/tmp/{{user-input}}|../../etc/passwd|/tmp/../../etc/passwd|/etc/passwd|Yes|
|12|/tmp/{{user-input}}| |/tmp/|/tmp|No|
|13|/tmp/{{user-input}}|../tmp|/tmp/../tmp|/tmp|Yes|
|14|/tmp/{{user-input}}|../tmp/|/tmp/../tmp/|/tmp|Yes|
|15|/tmp/{{user-input}}|../tmp-lookalike|/tmp/../tmp-lookalike|/tmp-lookalike|Yes|
|16|/tmp/{{user-input}}|../tmp/abcd|/tmp/../tmp/abcd|/tmp/abcd|No|
|17|/tmp/{{user-input}}|..artifact.txt|/tmp/..artifact.txt|/tmp/..artifact.txt|No|
|18|/tmp/{{user-input}}|artif..t.txt|/tmp/artif..t.txt|/tmp/artif..t.txt|No|
|19|/tmp/subdir/{{user-input}}|../../tmp/subdir/|/tmp/subdir/../../tmp/subdir/|/tmp/subdir|Yes|
|20|/tmp/subdir/{{user-input}}|../../tmp/subdir/abcd|/tmp/subdir/../../tmp/subdir/abcd|/tmp/subdir/abcd|No|

**Comment by alexec** [comment]
> Here is a table showing the cases that this code is tested for. If any of these cases does not make sense, or if I'm missing a case, please let me know.

I'm afraid I've not got bandwidth can't provide any detailed feedback on this. Suffice to say, you only need to find one counterexample to show this is not a complete list.

Examples 4..7 are interesting. I'm not sure what they'd mean and if we need to support them. What happens if these become error cases?

What I'm pushing for is `CleanPath()` to be simplified.

**Review by alexec — CHANGES_REQUESTED** [review]
@kennytrytek we've gotten to an impass here. I'm won't approve any changes that I deem will create a maintanance burden, such as regex path matching, yet I'd still like to address this risk. I'd prefer to just reject ".." in artifact path names. Let me know if you're still interested in submitting that as a PR.

**Comment by kennytrytek** [comment]
> such as regex path matching

I still maintain that I am not using regex path matching in this PR, but using regex to define "/.." and "/../" strings in an OS-agnostic way. I could do the same operation without using the regex library, but it was simpler to reason about using regex tools.

No, I would not like to submit ".." path checking, since it does not solve the problem, as discussed a few times already. If this pull request will not be accepted, please close it.

**Comment by crenshaw-dev** [comment]
> I could do the same operation without using the regex library

Agreed, the regex is very simple and could be replaced with strings.EndsWith (or whatever the golang function for that operation is).

The logic looks really robust. But I'm sympathetic to the idea that we might miss some edge case, and we'll encounter maintenance cost.

Personally, I'm in favor of this PR (maybe with the regex subbed for a simpler utility). But if we went with a simple `..` check, I think we'd want to feature-flag it. I can certainly imagine someone using double-dots in filenames, and we should avoid making their lives difficult. Of course, even a feature-flagged "simple" check brings its own maintenance cost.

**Comment by alexec** [comment]
Do we want a feature flag in this PR?

**Comment by crenshaw-dev** [comment]
@alexec I'd totally be in favor of that. I trust the unit tests enough to default it to true, but even default-false would be a big improvement over no check possible at all.

**Comment by crenshaw-dev** [comment]
Also @kennytrytek your work on this is really appreciated. I think any frustration is due to the intractability of the problem and not your solution. This PR and the comments are exceedingly thorough. I wish all my work were this quality. :-)

**Review by alexec — APPROVED** [review]
